Source/Pawnmorphs/Esoteria/UserInterface/Settings/Dialog_BlacklistAnimal.cs

```diff
@@ -0,0 +1,177 @@
+			curY += 35;
+
+			float totalHeight = inRect.height - curY - Math.Max(APPLY_BUTTON_SIZE.y, Math.Max(RESET_BUTTON_SIZE.y, CANCEL_BUTTON_SIZE.y)) - SPACER_SIZE;
+			_animalListBox.Draw(inRect, 0, curY, totalHeight, (item, listing) =>
+			{
+				bool current = _blacklistedAnimalDictonary.TryGetValue(item, false);
+				listing.CheckboxLabeled(item.LabelCap, ref current, item.modContentPack.ModMetaData.Name);
+				_blacklistedAnimalDictonary[item] = current;
+			});
+
+			// Draw the apply, reset and cancel buttons.
+			if (Widgets.ButtonImage(new Rect(5, inRect.height - 33, RAW_BUTTON_SIZE.x, RAW_BUTTON_SIZE.y), TexButton.Rename))
+			{
+				ShowRawInput();
+			}
+
+			float buttonVertPos = inRect.height - Math.Max(APPLY_BUTTON_SIZE.y, Math.Max(RESET_BUTTON_SIZE.y, CANCEL_BUTTON_SIZE.y));
+			float applyHorPos = inRect.width / 2 - APPLY_BUTTON_SIZE.x - RESET_BUTTON_SIZE.x / 2 - SPACER_SIZE;
+			float resetHorPos = inRect.width / 2 - RESET_BUTTON_SIZE.x / 2;
+			float cancelHorPos = inRect.width / 2 + RESET_BUTTON_SIZE.x / 2 + SPACER_SIZE;
+			if (Widgets.ButtonText(new Rect(applyHorPos, buttonVertPos, APPLY_BUTTON_SIZE.x, APPLY_BUTTON_SIZE.y), APPLY_BUTTON_TEXT))
+			{
+				OnAcceptKeyPressed();
+			}
+			if (Widgets.ButtonText(new Rect(resetHorPos, buttonVertPos, RESET_BUTTON_SIZE.x, RESET_BUTTON_SIZE.y), RESET_BUTTON_TEXT))
+			{
+				RimWorld.SoundDefOf.Checkbox_TurnedOff.PlayOneShotOnCamera(null);
+				RefreshList();
+			}
+			if (Widgets.ButtonText(new Rect(cancelHorPos, buttonVertPos, CANCEL_BUTTON_SIZE.x, CANCEL_BUTTON_SIZE.y), CANCEL_BUTTON_TEXT))
+			{
+				OnCancelKeyPressed();
+			}
+		}
+
+
+		private void ShowRawInput()
+		{
+			List<string> exportCollection = new List<string>();
+			exportCollection.AddRange(_blacklistedAnimalDictonary.Where(x => x.Value == false).Select(x => x.Key.defName).ToList());
+			exportCollection = exportCollection.Distinct().ToList();
+
+			string stringValue = string.Join(",", exportCollection);
+			Dialog_Textbox textBox = new Dialog_Textbox(stringValue, true, new Vector2(300, 100));
+
+			textBox.ApplyAction = (value) =>
+			{
+				try
+				{
+					_settingsReference.Clear();
+					_settingsReference.AddRange(value.Split(',').Where(x => String.IsNullOrWhiteSpace(x) == false).Distinct());
+					RefreshList();
+				}
+				catch (Exception)
+				{
+					Find.WindowStack.Add(new Dialog_Popup(FAILED_PARSING_TEXT, new Vector2(300, 100)));
+				}
+			};
+
+			Find.WindowStack.Add(textBox);
+		}
+
+		private void ApplyChanges()
+		{
+			// We want to persist any definitions assigned for defnames that could not be loaded into the game at present time.
+			// Like animals from currently disabled mods.
+			// Take the old list, add all disabled animals (this will cause duplicates) and then apply distinct.
+			_settingsReference.AddRange(_blacklistedAnimalDictonary.Where(x => x.Value == false).Select(x => x.Key.defName).ToList());
+			_settingsReference = _settingsReference.Distinct().ToList();
```

Comment:
I don't think this will work right for a few reasons:
- Un-blacklisting an animal won't actually remove it from the blacklist, because .AddRange().Distinct() only cares about newly-blacklisted animals.
- Calling _settingsReference = _settingsReference.Distinct().ToList() doesn't edit the original collection from the settings window, so that part doesn't get persisted.  (_settingsReference should probably be declared readonly to ensure you don't accidentally replace the reference)

I'm trying to think of a way to think of a better way to do it.  The easiest would probably be to use intermediate iterables.  Something like this:

```
var existingBlacklist = _settingsReference.ToHashSet(); 
var blacklist = _blacklistedAnimalDictonary.Where(x => x.Value == false).Select(x => x.Key.defName);
var whitelist = _blacklistedAnimalDictonary.Where(x => x.Value == true).Select(x => x.Key.defName);

_settingsReference.Clear();
// Ensure existing blacklist entries are only removed if they are explicitly un-blacklisted
// That way defs that may not exist (i.e. because they're from an unloaded mod) will be persisted
_settingsReference.AddRange(existingBlacklist.Union(blacklist).Except(whitelist));
```

Reply:
It was supposed to have a clear function in there.

But i also don't want to remove defs that may not currently be loaded.

Say i add genetic rim, ban a pile of animals, then decide to do a game with a mod replacing it, and then banning a number of animals in _that_.

I don't want to loose all the defs i defined in genetic rim.